libraries/AP_NavEKF2/AP_NavEKF2_core.cpp

```diff
@@ -1511,4 +1521,38 @@ void NavEKF2_core::zeroAttCovOnly()
     }
 }
 
+/*
+  setup earth field state from full declination, inclination and intensity tables
+  Called when we get 3D GPS lock if enabled
+ */
+void NavEKF2_core::setEarthFieldFromTables(void)
+{
+    float field_intensity, declination_deg, inclination_deg;
+    const Location &loc = _ahrs->get_gps().location();
+    if (!AP_Declination::get_mag_field_ef(loc.lat*1.0e-7, loc.lng*1.0e-7,
+                                          field_intensity, declination_deg, inclination_deg)) {
+        return;
+    }
+
+    /*
+      we are initialising the full earth field using the tables
+    */
+    Vector3f mag_ef(field_intensity, 0.0f, 0.0f);
+    Matrix3f R;
+    R.from_euler(0.0f, -ToRad(inclination_deg), ToRad(declination_deg));
+    mag_ef = R * mag_ef;
+    stateStruct.earth_magfield = mag_ef;
+
+    /*
+      set small initial variances for the earth field states
+    */
+    setDiagonal(P,16,18,1.0e-6);
```

Comment:
We also need to handle the body field states. I would recommend also zeroing row and columns 19...21 and setting the corresponding diagonals to twice the magnetometer observation noise squared. The body field states should also be zeroed.

Reply:
only zero 19..21 cols/rows? Why not 16..18 as well?

Reply:
You were already doing 16..18 :)